GAE/src/org/waterforpeople/mapping/app/web/RawDataRestServlet.java

```diff
@@ -143,35 +144,64 @@ protected RestResponse handleRequest(RestRequest req) throws Exception {
                 return null;
             }
 
-            // retrieve and update answers for current instance
-            Map<Long, QuestionAnswerStore> existingAnswers = qasDao.mapByQuestionId(qasDao
-                    .listBySurveyInstance(instance.getKey().getId()));
-            Map<Long, String[]> incomingAnswers = importReq.getQuestionAnswerMap();
+            // questionId -> iteration -> QAS
+            Map<Long, Map<Integer, QuestionAnswerStore>> existingAnswers = qasDao
+                    .mapByQuestionIdAndIteration(qasDao.listBySurveyInstance(instance.getKey()
+                            .getId()));
+
+            Map<Long, Map<Integer, String[]>> incomingResponses = importReq.getResponseMap();
+
             List<QuestionAnswerStore> updatedAnswers = new ArrayList<QuestionAnswerStore>();
 
-            for (Map.Entry<Long, String[]> qasEntry : incomingAnswers.entrySet()) {
-                QuestionAnswerStore answer = null;
-                if (existingAnswers.containsKey(qasEntry.getKey())) {
-                    answer = existingAnswers.get(qasEntry.getKey());
-                } else {
-                    answer = new QuestionAnswerStore();
-                    answer.setQuestionID(qasEntry.getKey().toString());
-                    answer.setSurveyInstanceId(instance.getKey().getId());
-                    answer.setSurveyId(s.getKey().getId());
-                    answer.setCollectionDate(instance.getCollectionDate());
+            for (Entry<Long, Map<Integer, String[]>> responseEntry : incomingResponses
+                    .entrySet()) {
+                Long questionId = responseEntry.getKey();
+                Map<Integer, String[]> iterationMap = responseEntry.getValue();
+
+                for (Entry<Integer, String[]> iterationEntry : iterationMap.entrySet()) {
+                    Integer iteration = iterationEntry.getKey();
+                    String response = iterationEntry.getValue()[0];
+                    String type = iterationEntry.getValue()[1];
+                    QuestionAnswerStore answer = null;
+                    if (existingAnswers.containsKey(questionId)) {
+                        if (existingAnswers.get(questionId).containsKey(iteration)) {
+                            answer = existingAnswers.get(questionId).get(iteration);
+                        }
+                    }
+
+                    // New answer/iteration
+                    if (answer == null) {
+                        answer = new QuestionAnswerStore();
+                        answer.setQuestionID(questionId.toString());
+                        answer.setSurveyId(instance.getKey().getId());
+                        answer.setSurveyId(s.getKey().getId());
```

Comment:
Why is the survey ID set up twice? Isn't the second call just overriding the first value?